nv_peer_mem.c

```diff
@@ -39,18 +39,67 @@
 #include <linux/export.h>
 #include <linux/hugetlb.h>
 #include <linux/atomic.h>
+#include <linux/pci.h>
 
 
 #include "nv-p2p.h"
 #include <rdma/peer_mem.h>
 
 
 #define DRV_NAME	"nv_mem"
-#define DRV_VERSION	"1.0"
+#define DRV_VERSION	"1.1"
 #define DRV_RELDATE	__DATE__
 
-#define peer_err(FMT, ARGS...) printk(KERN_ERR   DRV_NAME " %s:%d " FMT, __FUNCTION__, __LINE__, ## ARGS)
+#define peer_err(FMT, ARGS...) printk(KERN_ERR   DRV_NAME " %s:%d " FMT, __func__, __LINE__, ## ARGS)
 
+#define INFO_ENABLE 1
+#define peer_info(FMT, ARGS...)		\
+	do {				\
+		if (INFO_ENABLE)		\
+			printk(KERN_INFO   DRV_NAME " %s:%d " FMT, __func__, __LINE__, ## ARGS);	\
+	} while (0)
+
+#ifndef NVIDIA_P2P_MAJOR_VERSION_MASK
+#define NVIDIA_P2P_MAJOR_VERSION_MASK   0xffff0000
+#endif
+
+#ifndef NVIDIA_P2P_MINOR_VERSION_MASK
+#define NVIDIA_P2P_MINOR_VERSION_MASK   0x0000ffff
+#endif
+
+#ifndef NVIDIA_P2P_MAJOR_VERSION
+#define NVIDIA_P2P_MAJOR_VERSION(v)	\
+	(((v) & NVIDIA_P2P_MAJOR_VERSION_MASK) >> 16)
+#endif
+
+#ifndef NVIDIA_P2P_MINOR_VERSION
+#define NVIDIA_P2P_MINOR_VERSION(v)	\
+	(((v) & NVIDIA_P2P_MINOR_VERSION_MASK))
+#endif
+
+/*
+ *	Note: before major version 2, struct dma_mapping had no version field,
+ *	so it is not possible to check version compatibility. In this case
+ *	let us just avoid dma mappings altogether.
+ */
+#if defined(NVIDIA_P2P_DMA_MAPPING_VERSION) &&	\
+	(NVIDIA_P2P_MAJOR_VERSION(NVIDIA_P2P_DMA_MAPPING_VERSION) >= 2)
```

Comment:
Who defines `NVIDIA_P2P_DMA_MAPPING_VERSION`? Is it in the nv-p2p.h header file?